feat(backends/s3): add warmup support before repacks and restores

gilbsgilbs opened the pull request:
<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.
-->

What does this PR change? What problem does it solve?
-----------------------------------------------------

This commit introduces basic support for transitioning pack files stored in cold storage to hot storage on S3 and S3-compatible providers.

To prevent unexpected behavior for existing users, the feature is gated behind new flags:

- `s3.enable-restore`: opt-in flag (defaults to false)
- `s3.restore-days`: number of days for the restored objects to remain in hot storage (defaults to `7`)
- `s3.restore-timeout`: maximum time to wait for a single restoration (default to `1 day`)
- `s3.restore-tier`: retrieval tier at which the restore will be processed. (default to `Standard`)

As restoration times can be lengthy, this implementation preemptively restores selected packs to prevent incessant restore-delays during downloads. This is slightly sub-optimal as we could process packs out-of-order (as soon as they're transitioned), but this would really add too much complexity for a marginal gain in speed.

To maintain simplicity and prevent resources exhautions with lots of packs, no new concurrency mechanisms or goroutines were added. This just hooks gracefully into the existing routines.

**Limitations:**

- Tests against the backend were not written due to the lack of cold storage class support in MinIO. Testing was done manually on Scaleway's S3-compatible object storage and AWS S3. If necessary, automatic testing with LocalStack or mocks can be explored, though this requires further discussion.
- Currently, this feature only warms up before restores and repacks (prune/copy), as those are the two main use-cases I came across. Support for other commands may be added in future iterations, as long as affected packs can be calculated in advance.
- The feature is gated behind a new alpha `s3-restore` feature flag to make it explicit that the feature is still wet behind the ears.
- There is no explicit user notification for ongoing pack restorations. While I think it is not necessary because of the opt-in flag, showing some notice may improve usability (but would probably require major refactoring in the progress bar which I didn't want to start). Another possibility would be to add a flag to send restores requests and fail early.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

- https://github.com/restic/restic/issues/3202 (probably doesn't technically close it as some commands may still not be supported)
- https://github.com/restic/restic/issues/2504

Checklist
---------

- [x] I have added tests for all code changes.
  - see "Limitations" section above: tests have only been written for warmups, but not for actual S3-backend restores because MinIO server doesn't support cold storage classes. I'm happy to discuss on the best way to write tests for that without adding too much boilerplate (e.g. introduce localstack or find a proper way to mock this).
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I'm done! This pull request is ready for review.

gilbsgilbs committed:
feat(backends/s3): add warmup support before repacks and restores

This commit introduces basic support for transitioning pack files stored
in cold storage to hot storage on S3 and S3-compatible providers.

To prevent unexpected behavior for existing users, the feature is gated
behind new flags:

- `s3.enable-restore`: opt-in flag (defaults to false)
- `s3.restore-days`: number of days for the restored objects to remain
  in hot storage (defaults to `7`)
- `s3.restore-timeout`: maximum time to wait for a single restoration
  (default to `1 day`)
- `s3.restore-tier`: retrieval tier at which the restore will be
  processed. (default to `Standard`)

As restoration times can be lengthy, this implementation preemptively
restores selected packs to prevent incessant restore-delays during
downloads. This is slightly sub-optimal as we could process packs
out-of-order (as soon as they're transitioned), but this would really
add too much complexity for a marginal gain in speed.

To maintain simplicity and prevent resources exhautions with lots of
packs, no new concurrency mechanisms or goroutines were added. This just
hooks gracefully into the existing routines.

**Limitations:**

- Tests against the backend were not written due to the lack of cold
  storage class support in MinIO. Testing was done manually on
  Scaleway's S3-compatible object storage. If necessary, we could
  explore testing with LocalStack or mocks, though this requires further
  discussion.
- Currently, this feature only warms up before restores and repacks
  (prune/copy), as those are the two main use-cases I came across.
  Support for other commands may be added in future iterations, as long
  as affected packs can be calculated in advance.
- The feature is gated behind a new alpha `s3-restore` feature flag to
  make it explicit that the feature is still wet behind the ears.
- There is no explicit user notification for ongoing pack restorations.
  While I think it is not necessary because of the opt-in flag, showing
  some notice may improve usability (but would probably require major
  refactoring in the progress bar which I didn't want to start). Another
  possibility would be to add a flag to send restores requests and fail
  early.

See https://github.com/restic/restic/issues/3202

MichaelEischer commented on internal/repository/warmup.go:
```diff
@@ -0,0 +1,109 @@
+type PacksWarmer struct {
+	repo        restic.Repository
+	packs       restic.IDSet
+	packsResult map[restic.ID]error
```
I'd prefer to use a sync.Map here. That should clean up the map accesses quite a bit.

gilbsgilbs replied:
Well, the sync.Map implementation really looks worse in practice.

- `sync.Map` loses type safety
- because of this, a type assertion is needed in getResult to cast `interface{}` to `error` + a null-assertion in case the error is nil
- a mutex is still used for the initial packs registration. (Well, the mutex could technically be removed because no operation is concurrent in practice, but doing `if !set.Has(x) { set.Insert(x, y) }` without a lock screams future problems)

gilbsgilbs replied:
I ended up removing this struct entirely. Resolving.

MichaelEischer reviewed (changes requested):
The whole approach still feels somewhat experimental, so I think we should move the code behind an alpha feature flag.

In general, automatically warming up and waiting for the required pack files looks like a solid approach. I prefer that variant compared to having `--warmup` flags that require the user to guess when the warmup is complete (and in case of prune won't interact well with concurrent modifications of the repository). However, to further progress cold storage support, we'll need at least some feedback for the user to indicate that restic is actually waiting for files to be warmed up in contrast to being stuck.

Having more supported commands would also be nice, but that can be changed later on while the feature is still in alpha state.

gilbsgilbs commented on internal/backend/s3/s3.go:
```diff
@@ -445,3 +450,99 @@ func (be *Backend) Delete(ctx context.Context) error {
+	}
+	opts.SetGlacierJobParameters(minio.GlacierJobParameters{Tier: minio.TierType(be.cfg.RestoreTier)})
+
+	err = be.client.RestoreObject(ctx, be.cfg.Bucket, filename, "", opts)
```
> I guess there's no API to restore multiple objects at once?

Not that I know of. Business is business.

> How expensive are those API calls when a file is already restore / has to be restored?

AWS being AWS, I assume they consider this a customer-initiated POST request like any other, so a few cts per 1000 request depending on the storage class: https://aws.amazon.com/s3/pricing/ .

If we're going have to restore almost all files, then it's cheaper to just restore them dumbly as I did. Otherwise, it'd be cheaper to HEAD first, then restore only when cold and no restore is pending. There's a third (cheaper) approach of listing all the files in the bucket first, and determining their restore status using the [`x-amz-optional-object-attributes: RestoreStatus`](https://aws.amazon.com/about-aws/whats-new/2023/06/amazon-s3-restore-status-s3-glacier-objects-list-api/) header, however I'm not sure how widely adopted this is among S3-compatible providers.

gilbsgilbs replied:
> I guess there's no API to restore multiple objects at once?

Not that I know of. Business is business.

> How expensive are those API calls when a file is already restore / has to be restored?

AWS being AWS, I assume they consider this a customer-initiated POST request like any other, so a few cts per 1000 request depending on the storage class: https://aws.amazon.com/s3/pricing/ .

If we're going to restore almost all files, then it's cheaper to just restore them dumbly as I did. Otherwise, it'd be cheaper to HEAD first to get the restore status, then restore only when cold and no restore is pending. There's a third (cheaper) approach of listing all the files in the bucket first, and determining their restore status using the [`x-amz-optional-object-attributes: RestoreStatus`](https://aws.amazon.com/about-aws/whats-new/2023/06/amazon-s3-restore-status-s3-glacier-objects-list-api/) header, however I'm not sure how widely adopted this is among S3-compatible providers.

For non-restore commands, the cost is probably negligible as you'll only restore a few packs anyway. To restore a whole repo, I'm not so sure if those requests would weight a lot compared the actual cost of restoring the data (which I'd expect to be abysmal).

Handling batch operations in the backend itself definitely adds some complexity, but I remember going back-and-forth on this subject when implementing this PR.

MichaelEischer replied:
My guess would be that most restores are partial restores. But that will vary depending on the exact use case.  Calling HEAD first seems reasonable tough.

gilbsgilbs commented on internal/backend/s3/s3.go:
```diff
@@ -445,3 +450,99 @@ func (be *Backend) Delete(ctx context.Context) error {
+	if be.cfg.RestoreDays != 0 {
+		opts.SetDays(be.cfg.RestoreDays)
+	}
+	opts.SetGlacierJobParameters(minio.GlacierJobParameters{Tier: minio.TierType(be.cfg.RestoreTier)})
```
They don't specify both in their examples, so I presume not: https://github.com/minio/minio-go/blob/8b220d8e58cc4e1bbb667b732a130c7b647d8db9/docs/API.md?plain=1#L1190 https://github.com/minio/minio-go/blob/8b220d8e58cc4e1bbb667b732a130c7b647d8db9/examples/s3/restoreobject.go#L51

AWS isn't specific about any behavioral difference: https://docs.aws.amazon.com/AmazonS3/latest/API/API_RestoreObject.html#AmazonS3-RestoreObject-request-GlacierJobParameters . I can just notice a `Tier` is required in `GlacierJobParameters` and not in `RestoreObjectRequest`, so maybe that's just for this.

MichaelEischer replied:
Ok, then let's start with the current approach and see whether someone complains :-) .

gilbsgilbs commented on internal/backend/s3/s3.go:
```diff
@@ -445,3 +450,99 @@ func (be *Backend) Delete(ctx context.Context) error {
+			return err
+		}
+
+		storageClass := objectInfo.Metadata.Get("X-Amz-Storage-Class")
```
Yup, that's a shame. IIRC, it's only filled on LIST operations when specifying the `x-amz-optional-object-attributes: RestoreStatus` header in the request. But as I said above, I'm not sure how widely implemented this is. I think this will work reliably, even though I admit I'm not too happy about it either.

gilbsgilbs commented on internal/backend/s3/s3.go:
```diff
@@ -445,3 +450,99 @@ func (be *Backend) Delete(ctx context.Context) error {
+		}
+
+		select {
+		case <-time.After(1 * time.Minute):
```
The only way I can imagine for an "event-based" approach would be through SNS, so definitely not "simple". Polling looks way simpler and more reliable to me. We only poll one pack at a time, so it shouldn't cost too much either. And we expect restores to be long, so one minute late at worse seems acceptable to me.

MichaelEischer replied:
Fine with me.

gilbsgilbs commented on doc/faq.rst:
```diff
@@ -242,3 +242,31 @@ collect a list of all files, causing the following error:
+- Restic will prevent sending metadata files (such as config files, lock files
+  or tree blobs) to Glacier or Deep Archive. Standard class is used instead to
+  ensure normal and fast operations for most tasks.
+- Currently, only the following commands are known to work or have worked:
```
> I think we should add an alpha feature flag to make it clear that there is no full blown cold storage support.

I'm all for making it 100% clear that the feature is experimental to the user, so I agree with the idea. But could you be more specific about what you mean by "an alpha feature flag"? What would it look like? The current flag is already 100% opt-in, so would you just rename it to "s3.enable-experimental-restores" or something like that?

gilbsgilbs commented:
Thanks @MichaelEischer for the feedback. I'm happy that you find this approach reasonable and that we can make cold storage support move forward in restic.

- I think I've responded to your questions. This PR and the choices I made when writing it are a bit cold in my mind (pun not intended) and I'm dealing with other personal struggles, so sorry if I am not perfectly on point. I'll see if I can do anything pragmatic to close any debate on "slightly oriented" questions, but tell me if you see any blocker.
- I either agree or don't have strong opinions on your other non-question comments.

So I'll rebase my branch and fix everything probably tomorrow, or by the end of next week.

> However, to further progress cold storage support, we'll need at least some feedback for the user to indicate that restic is actually waiting for files to be warmed up in contrast to being stuck.

IIRC, this is more tedious than it seems because the current progress bar implementationS are "swallowing" everything. I'll see what I can do (probably in another branch if you don't mind, so as to not pollute this PR with UI-related stuff).

gilbsgilbs commented on internal/backend/s3/s3.go:
```diff
@@ -445,3 +450,99 @@ func (be *Backend) Delete(ctx context.Context) error {
+		}
+
+		storageClass := objectInfo.Metadata.Get("X-Amz-Storage-Class")
+		if !slices.Contains(archiveClasses, storageClass) {
```
I guess it could make sense to abort early if a restore is not ongoing, so that we don't wait for the timeout. Something like this:


```go
		storageClass := objectInfo.Metadata.Get("X-Amz-Storage-Class")
		if !slices.Contains(archiveClasses, storageClass) {
			debug.Log("s3 file is restored: %s\n", filename)
			return nil
		}
		if objectInfo.Restore != nil && objectInfo.Restore.OngoingRestore == false {
			return errors.New("s3 restore is not ongoing")
		}
```

However, I can also imagine cases where it could cause unwanted failures depending on how the S3 implementation is done. For example, if `x-amz-storage-class` is not updated perfectly in sync with the `x-amz-restore` header or if the `x-amz-restore` header is not set just after the restore request, it could fail for nothing. I'm not sure we want to be that rigid (or at least I'm not very confident it will work as reliably and predictably).

gilbsgilbs committed:
ui: warn user when files are warming up from cold storage

gilbsgilbs committed:
refactor: remove the PacksWarmer struct

It's easier to handle multiple handles in the backend directly, and it
may open the door to reducing the number of requests made to the backend
in the future.

gilbsgilbs commented:
I just removed the `PacksWarmer` struct which makes my whole implementation way simpler by removing one layer of indirection. We can always add such structure in the future if it's needed. Giving all handles to the backend at once may also allow optimizing the number of requests sent (which I have not implemented but is now theoretically possible).

gilbsgilbs committed:
fix(backends/s3): fix restores on AWS S3

MichaelEischer reviewed (changes requested):
Thanks for the changes so far. I have a few more comments below. Regarding the progress information, let's stick to the simple log messages for now. I don't see a simple way to add proper progress bars without a lot of refactoring beforehand.

gilbsgilbs committed:
fix: lint

gilbsgilbs commented on internal/repository/repack.go:
```diff
@@ -18,16 +18,30 @@ type repackBlobSet interface {
+	dstRepo restic.Repository,
+	packs restic.IDSet,
+	keepBlobs repackBlobSet,
+	p *progress.Counter,
```
I just want to point out that a progress bar for restores would not add a lot of value because all restores will succeed approximately at the same time. A progress bar would remain at "0%" for hours, and then suddenly jump to 90-99-100%. AFAIK, AWS doesn't provide any ETA for restores either. So I don't think heavy works are justified at all for this feature.

gilbsgilbs commented on internal/backend/s3/s3.go:
```diff
@@ -440,3 +449,140 @@ func (be *Backend) Delete(ctx context.Context) error {
+		if restore.OngoingRestore {
+			return false, true
+		}
+		if !restore.ExpiryTime.IsZero() {
```
I wasn't sure if calling `RestoreObject` on an already-restored object "renews" the lease or is just ignored, so I didn't want to write any explicit logic about that. I need to check if it is possible to do this.

Also, even if it worked, I am not sure how to define "expire soon". If it's `< RestoreDays`, then we may just call `RestoreObject` every time and not bother. If it's less, how much? `< RestoreDays / 2` ? Sounds arbitrary. Letting user configure this value sounds too specific.

gilbsgilbs replied:
Just tested it and it seems to work. `RestoreDays` can be increased as well as reduced with subsequent calls to `RestoreObject`.

Given that the `expiry-date` is rounded up to the next day (GMT) by AWS, I think it is safe to call `RestoreObject` again if the current `expiry-date` is set before `now + RestoreDays`. This guarantees `RestoreDays` is always respected (at least on AWS), and still avoids calling `RestoreObject` for nothing.

ca5f6d5be42196b05b3d2073af4a09af245f27a7

gilbsgilbs committed:
fix: address review comments

gilbsgilbs commented on internal/backend/s3/s3.go:
```diff
@@ -440,3 +449,140 @@ func (be *Backend) Delete(ctx context.Context) error {
+}
+
+// getWarmupStatus returns whether the given objectInfo is warm or warming up.
+func (be *Backend) getWarmupStatus(objectInfo minio.ObjectInfo) (isWarm bool, isWarmingUp bool) {
```
Sorry, I was just lazy because I was annoyed that AWS S3 and Scaleway Object Storage were behaving differently. An enum also prevents inconsistent states such as `isWarm=true isWarmingUp=true`, so yes, that's inherently better. I moved to an enum in the latest changeset. Thanks.

gilbsgilbs committed:
fix(backends/s3): resend restore-object when approaching expiry-time

MichaelEischer committed:
fix typos

MichaelEischer reviewed (approved):
LGTM. Thanks for the quick changes! I've added a tiny fixup commit to fix a typo and correctly use `feature.TestSetFlag()`.

MichaelEischer committed:
fix linter warning

gilbsgilbs commented:
I somehow didn't notice this was merged. Thank you very much for your time @MichaelEischer . I'll follow-up with the missing commands in a few days or weeks.
